ic-os/components/defs.bzl

```diff
@@ -0,0 +1,48 @@
+"""
+Rules for check unused components checks.
+"""
+
+def _check_unused_components_test_impl(ctx):
+    repo_components = [component.label.name for component in ctx.attr.repo_components]
+    used_components = [component.label.name for component in ctx.attr.used_components]
+
+    # ADDITIONAL_USED_COMPONENT_FILES are files used for testing and development
+    ADDITIONAL_USED_COMPONENT_FILES = ["networking/dev-certs/canister_http_test_ca.key", "networking/dev-certs/root_cert_gen.sh"]
+    used_components += ADDITIONAL_USED_COMPONENT_FILES
+
+    unused_components = [file for file in repo_components if file not in used_components]
+
+    unused_components_file = ctx.actions.declare_file(ctx.label.name + ".unused_components")
+    ctx.actions.write(
+        output = unused_components_file,
+        content = "\n".join(unused_components),
+    )
+
+    if unused_components:
+        fail("Unused components found: {}".format(", ".join(unused_components)))
+
+    script = ctx.actions.declare_file(ctx.label.name + ".sh")
+    ctx.actions.write(
+        output = script,
+        content = "echo Unused components check completed",
+    )
```

Comment:
Bazel demanded the test have an executable script. Rather than move any logic into the executable, I found it simpler to make it as dumb as possible, but I'm sure there's a more bazel (and likely better) way to do this...